### ISSUE TITLE
Reproducible init of trainable variables (TVs) even when number of TVs changes

### ISSUE THREAD

**Issue by edwardyehuang** [tensorflow/tensorflow#54442]
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow):Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): CentOS
- TensorFlow installed from (source or binary): PIP
- TensorFlow version (use command below):2.8
- Python version:3.8.12
- GPU model and memory:8×v100

For example:

Model has 3 vars, varA, varB and varC

Now, we change the model with a new varA2, insert between varA and varB. So the model has 4 vars, varA, varA2, varB and varC.

Even varA2 is not used except weight initialization, e.g. tf.keras.initializers.GlorotUniform(). The model will output different result than before, because now the varA2 may use the weights of pervious varB and so on.

It is possible to binding the weight random initialization with its name?

@duncanriach 

https://colab.research.google.com/drive/19Ui9Fp1JF06Xl440pnHZ8zis-yyTx9jq?usp=sharing

**Comment by edwardyehuang** [comment]
@mohantym  I think it is a bug

**Comment by mohantym** [comment]
Ok @edwardyehuang ! Did you check this[ thread ](https://github.com/tensorflow/tensorflow/issues/36650#issuecomment-863186242)on using tf.name_scope on the targeted weights though?

**Comment by mohantym** [comment]
Ok!  Changing the labels then. But Can you please update the template with standalone code to reproduce this issue? Thanks!

**Comment by edwardyehuang** [comment]
I will try to reproduce in colab, currently colab seems not support TensorFlow 2.8

**Comment by mohantym** [comment]
It does support TF 2.8 now. Thank you!

**Comment by edwardyehuang** [comment]
https://colab.research.google.com/drive/19Ui9Fp1JF06Xl440pnHZ8zis-yyTx9jq?usp=sharing

**Comment by duncanriach** [comment]
This behavior is:
* not a bug.
* expected in all versions of TensorFlow.
* not related to GPU functionality.

There is no need to restart the runtime to reset the global pseudo-random number generator. You can simply call `tf.random.set_seed`.

```
for i in range(2):
  tf.random.set_seed(0)
  a = tf.Variable(tf.random_uniform_initializer()(shape=(1,)))
  c = tf.Variable(tf.random_uniform_initializer()(shape=(1,)))
  print("run: %d; a: %f; c: %f" % (i, a, c))

run: 0; a: -0.020802; c: 0.005541
run: 1; a: -0.020802; c: 0.005541
```

The following now demonstrates the behavior you are noticing. The global pseudo-random number generator will be in a different state after initializing `b`, and therefore `c` will be initialized differently than it was before. This is normal and expected behavior.

```
tf.random.set_seed(0)
a = tf.Variable(tf.random_uniform_initializer()(shape=(1,)))
b = tf.Variable(tf.random_uniform_initializer()(shape=(1,)))
c = tf.Variable(tf.random_uniform_initializer()(shape=(1,)))
print("run: %d; a: %f; b: %f; c: %f" % (2, a, b, c))

run: 2; a: -0.020802; b: 0.005541; c: -0.030479
```

If you would like each variable to be initialized using pseudo-random number generator state that is independent of the most recent state of the global pseudo-random number generator, then you can provide a `seed` parameter to the initializers used for the variables initialized after the change in the model (or to all initializers).

```
for i in range(2):
  tf.random.set_seed(0)
  a = tf.Variable(tf.random_uniform_initializer()(shape=(1,)))
  if (i == 1):
    b = tf.Variable(tf.random_uniform_initializer()(shape=(1,)))
  c = tf.Variable(tf.random_uniform_initializer(seed=42)(shape=(1,)))
  print("run: %d; a: %f; c: %f" % (i, a, c))
  
run: 0; a: -0.020802; c: -0.021757
run: 1; a: -0.020802; c: -0.021757
```

The above code can be run in [this colab notebook](https://colab.research.google.com/drive/1LX3FmX0Vdka14SnXuRpvpxioCrJxseLF?usp=sharing).

**Comment by duncanriach** [comment]
I also suggest the following for a more specific title:

Reproducible init of trainable variables (TVs) even when number of TVs changes

**Comment by edwardyehuang** [comment]
> This behavior:
> 
> * is not a bug,
> * is expected in all versions of TensorFlow,
> * is not related to GPU functionality, and
> * is not related to `tf.config.experimental.enable_op_determinism`.
> 
> First of all, there is no need to restart the runtime to reset the global pseudo-random number generator. You can simply call `tf.random.set_seed`.
> 
> ```
> for i in range(2):
>   tf.random.set_seed(0)
>   a = tf.Variable(tf.random_uniform_initializer()(shape=(1,)))
>   c = tf.Variable(tf.random_uniform_initializer()(shape=(1,)))
>   print("run: %d; a: %f; c: %f" % (i, a, c))
> 
> run: 0; a: -0.020802; c: 0.005541
> run: 1; a: -0.020802; c: 0.005541
> ```
> 
> Next, the following demonstrates the behavior you are noticing. The global pseudo-random number generator will be in a different state after initializing `b`, and therefore `c` will be initialized differently than it was before. This is normal and expected behavior.
> 
> ```
> tf.random.set_seed(0)
> a = tf.Variable(tf.random_uniform_initializer()(shape=(1,)))
> b = tf.Variable(tf.random_uniform_initializer()(shape=(1,)))
> c = tf.Variable(tf.random_uniform_initializer()(shape=(1,)))
> print("run: %d; a: %f; b: %f; c: %f" % (2, a, b, c))
> 
> run: 2; a: -0.020802; b: 0.005541; c: -0.030479
> ```
> 
> Finally, if you would like each variable to be initialized using pseudo-random number generator state that is independent of the most recent state of the global pseudo-random number generator, then you can provide a `seed` parameter to the initializers used for the variables initialized after the change in the model (or to all initializers).
> 
> ```
> for i in range(2):
>   tf.random.set_seed(0)
>   a = tf.Variable(tf.random_uniform_initializer()(shape=(1,)))
>   if (i == 1):
>     b = tf.Variable(tf.random_uniform_initializer()(shape=(1,)))
>   c = tf.Variable(tf.random_uniform_initializer(seed=42)(shape=(1,)))
>   print("run: %d; a: %f; c: %f" % (i, a, c))
>   
> run: 0; a: -0.020802; c: -0.021757
> run: 1; a: -0.020802; c: -0.021757
> ```
> 
> The above code can be run in [this colab notebook](https://colab.research.google.com/drive/1LX3FmX0Vdka14SnXuRpvpxioCrJxseLF?usp=sharing).

Hi. I understand your workaround. However, I would like if this is automatically in franework-level, since this will be much easier to debug the performance of nerual network.

**Comment by duncanriach** [comment]
Using the `seed` parameter of the initializers is not a work-around, @edwardyehuang; there is definitely _**not**_ a bug that requires a work-around. For pseudo-random initialization of any given trainable variable to be non-variant under changes in the number and/or size of the trainable variables in the model (or any other changes in the use of the global pseudo-random number generator, PRNG), the `seed` parameters of the trainable variable initializers must be used.

This current issue is developing into a feature request. That would require the feature requirements to be clarified and an API to be defined.

-------

Here is a suggested proposal: when op determinism is expected (after [`tf.config.experimental.enable_op_determinism`](https://www.tensorflow.org/api_docs/python/tf/config/experimental/enable_op_determinism) has been called), the instantiation of any initializer without a `seed` parameter would throw an exception.

An argument in favor of this functionality could be something like: without having a `seed` parameter, the initializer will produce a different sequence of initialization values if the code-path changes such that the state of the global PRNG is different when the initializer is executed.

An argument against this functionality could be something like: we make no guarantee about the reproducibility of the overall program when the code-path is changed in any way. If you want this functionality, you can explicitly reset the PRNGs used in the initializers.

-------

This seems to be a new requirement entirely: the requirement that existing trainable variables are forced to initialize the same way even when the global PRNG is used differently elsewhere in the program. This goes beyond run-to-run determinism (with an unmodified program). I'm wondering how universally critical this requirement is and therefore whether this functionality should be forced (through exceptions) upon anyone who seeks only run-to-run determinism (with an unmodified program).

**Comment by duncanriach** [comment]
@edwardyehuang or @mohantym please will you change the tag from bug to feature.

Another potential functionality, when op determinism is expected, might be for initializers to assume a seed value of `0` if one is not provided via the `seed` parameter, or to generate a seed value from `name` if `name` is provided. The priority would be: use the `seed` parameter provided if one is provided, otherwise generate a `seed` from `name` if `name` is provided, else set `seed` to `0`.

I can't promise that any of this will be approved for the API, partly because I think it goes beyond the intended effects of [tf.config.experimental.enable_op_determinism](https://www.tensorflow.org/api_docs/python/tf/config/experimental/enable_op_determinism). I'm only trying to think through some possible ways this might work.

@edwardyehuang, can you provide more information about why this is an important feature, how it would be used, and give a sense of how commonly this feature might be needed by the community in general?

**Comment by chunduriv** [comment]
@edwardyehuang,

Please can you refer [comment1](https://github.com/tensorflow/tensorflow/issues/54442#issuecomment-1049458996), [comment2](https://github.com/tensorflow/tensorflow/issues/54442#issuecomment-1049526825) by @[duncanriach](https://github.com/duncanriach). 

Can you please provide more information  about why this is an important usecase? Thanks!

**Comment by edwardyehuang** [comment]
> @edwardyehuang,
> 
> Please can you refer [comment1](https://github.com/tensorflow/tensorflow/issues/54442#issuecomment-1049458996), [comment2](https://github.com/tensorflow/tensorflow/issues/54442#issuecomment-1049526825) by @[duncanriach](https://github.com/duncanriach).
> 
> Can you please provide more information about why this is an important usecase? Thanks!

Binding weights name with a fixed seed can help researcher figure out the extractly module/layer/design that helps to improve the overall performance:

For example, recently years sota publications often battle between pre-ln (layer normalization) or post-ln of Transformer:

> Pre-LN:
> 1. LayerNormalization
> 2. Token-mixer (e.g. Multi heads Self-Attention)

> Post-LN
> 1. Token-mixer (e.g. Multi heads Self-Attention)
> 2. LayerNormalization

That why the binding is important, since different order can have effect on seed if not specified.

BTW, after binding, we still can change the global seed to do more modifications.

**Comment by duncanriach** [comment]
The general principal behind run-to-run determinism with respect to data scientific exploration is that holding as many unrelated independent variables constant as possible helps to focus the experiment on the effect of the changes on the dependent variable. In other words, reducing unrelated noise increases the signal-to-noise ratio, and therefore the power of the experiment.

In this context, to paraphrase what you have written, when making a change to an architecture, the effect of that change is more easily determined when the initial state of the trainable variables is held constant when possible.

The example you have given is when the order of layers is changed. The original example you gave was equivalent to adding another later. I imagine that another example might be removing a layer, or changing the number of trainable variables in a layer.

This seems like a reasonable motive for the proposed feature, but let's remember that the required functionality is basically already available by setting the (relevant) initializer seeds; it only requires that the user knows about and understands the principles that we are discussing in this issue.

> BTW, after binding, we still can change the global seed to do more modifications.

I'll address this idea using a code example, which I have added to the end of [my colab notebook](https://colab.research.google.com/drive/1LX3FmX0Vdka14SnXuRpvpxioCrJxseLF?usp=sharing):

```
for i in range(2):
  tf.random.set_seed(i)
  print("global seed: %d" % i)
  for j in range(2):
    a = tf.Variable(tf.random_uniform_initializer(seed=10)(shape=(1,)))
    if (i == 1):
      b = tf.Variable(tf.random_uniform_initializer(seed=11)(shape=(1,)))
    c = tf.Variable(tf.random_uniform_initializer(seed=12)(shape=(1,)))
    print("run: %d; a: %f; c: %f" % (j, a, c))
    
global seed: 0
run: 0; a: -0.039600; c: 0.040857
run: 1; a: -0.039600; c: 0.040857
global seed: 1
run: 0; a: -0.039600; c: 0.040857
run: 1; a: -0.039600; c: 0.040857
```

This shows that when the initializer seeds are set, the global seed will, in fact, have no effect at all on the initializers. With the existing functionality, if we want a given initializer to follow the global seed, we just don't provide a local seed. With the proposed functionality, we would need to provide a local seed in order to override the automatically-generated local seed.

**Comment by duncanriach** [comment]
@reed, what are your thoughts on this proposed feature (see [this comment](https://github.com/tensorflow/tensorflow/issues/54442#issuecomment-1049526825)), which has motivation described in [this comment](https://github.com/tensorflow/tensorflow/issues/54442#issuecomment-1051545505)?

**Comment by sachinprasadhs** [comment]
The way random.set_seed works is in two different ways, `global seed` and `operation level seed`. 
You can see the complete details in the document here https://www.tensorflow.org/api_docs/python/tf/random/set_seed.
You can also set the seed for the initializers, for example for `GlorotNormal` [here](https://www.tensorflow.org/api_docs/python/tf/keras/initializers/GlorotNormal).

**Comment by edwardyehuang** [comment]
> The way random.set_seed works is in two different ways, `global seed` and `operation level seed`. You can see the complete details in the document here https://www.tensorflow.org/api_docs/python/tf/random/set_seed. You can also set the seed for the initializers, for example for `GlorotNormal` [here](https://www.tensorflow.org/api_docs/python/tf/keras/initializers/GlorotNormal).

Hi. I think you may overlooked discussion above.

**Comment by Venkat6871** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.
